cloudinit/distros/__init__.py

```diff
@@ -748,9 +851,13 @@ def _get_package_mirror_info(mirror_info, data_source=None,
         mirrors = []
         for tmpl in searchlist:
             try:
-                mirrors.append(tmpl % subst)
+                mirror = tmpl % subst
             except KeyError:
-                pass
+                continue
+
+            mirror = _sanitise_mirror_url(mirror)
+            if mirror is not None:
+                mirrors.append(mirror)
```

Comment:
In the case of non-transformable URLs, we just append the mirror as it is.  Can you comment on whether that's acceptable?  Is it worth logging/noting the difference here?

Reply:
In the case of a non-transformable URL, `_sanitise_mirror_url` would return `None`, so we wouldn't execute this line.

Perhaps this would be clearer if I didn't reuse the `mirror` variable (as below)? (Or perhaps I've misunderstood the question?)

```py
        for tmpl in searchlist:
            try:
                mirrors.append(tmpl % subst)
                mirror = tmpl % subst
            except KeyError:
                pass
                continue

            sanitised_mirror = _sanitise_mirror_url(mirror)
            if sanitised_mirror is not None:
                mirrors.append(sanitised_mirror)
```

Reply:
Previously we always appended the mirror.  I would think we want to keep doing that; the fix here is to apply a transformation if needed.  In the case of an "unparsable" url, this code no longer appends the mirror.  Is that expected behavior ?  I would have expected that we apply the transformation if possible but that we'd always append the mirror (transformed or in original form).

Reply:
This is a change in behaviour, you're correct. My assumption is that if Python's urlllib can't parse a generated URL, then apt won't be able to either. In which case, we're better off using a different pattern (or the configured fallback mirror) than the unparseable generated URL.

Note that Python's urllib uses a very permissive parser; if it can't handle a URL then it contains something strange like mismatched square brackets or weirdly-normalised Unicode characters.

Reply:
OK.  Let's call that out in the commit message.  I agree that since we're parsing it to sanitize it, we might as well filter out invalid mirror hostnames.  I'm resisting the urge to log this;  I think if we update docs around the mirror/apt to mention that we now will sanitize the mirror URLs and invalid mirror hosts will be dropped.  Documenting our new default behavior is enough to skip logging.